src/toxcore/io_tokio.rs

```diff
@@ -36,37 +36,40 @@ pub fn send_to_bounded<T: Send + 'static, Tx, E: Debug>(tx: &Tx, v: T, timeout:
 {
     Box::new(send_to(tx, v)
         .timeout(timeout)
-        .map_err(|e| Error::new(ErrorKind::Other,
+        .map_err(|e| IoError::new(IoErrorKind::Other,
             format!("Failed to send message: {}", e)
         )))
 }
 
 /// Send item to a sink using reference
-pub fn send_all_to<T: Send + 'static, S, Tx, E: Debug>(tx: &Tx, s: S) -> IoFuture<()>
+pub fn send_all_to<T: Send + 'static, S, Tx, E: Debug>(tx: &Tx, s: S) -> impl Future<Item=(), Error=IoError> + Send
     where S: Stream<Item = T, Error = E> + Send + 'static,
           Tx: Sink<SinkItem = T, SinkError = E> + Send + Clone + 'static
 {
-    Box::new(tx
+    tx
         .clone() // clone tx sender for 1 send only
         .send_all(s)
         .map(|_tx| ()) // ignore tx because it was cloned
         .map_err(|e| {
             // This may only happen if rx is gone
-            // So cast SendError<T> to a corresponding std::io::Error
             debug!("Send to a sink error {:?}", e);
-            Error::from(ErrorKind::UnexpectedEof)
+            IoError::new(IoErrorKind::Other, format!("{:?}",e))
         })
-    )
 }
 
 /// Send item to a sink using reference with timeout
-pub fn send_all_to_bounded<T: Send + 'static, S, Tx, E: Debug>(tx: &Tx, s: S, timeout: Duration) -> IoFuture<()>
+pub fn send_all_to_bounded<T: Send + 'static, S, Tx, E: Debug>(tx: &Tx, s: S, timeout: Duration) -> impl Future<Item=(), Error=IoError> + Send
     where S: Stream<Item = T, Error = E> + Send + 'static,
           Tx: Sink<SinkItem = T, SinkError = E> + Send + Clone + 'static
 {
-    Box::new(send_all_to(tx, s)
+    send_all_to(tx, s)
         .timeout(timeout)
-        .map_err(|e| Error::new(ErrorKind::Other,
-            format!("Failed to send message: {}", e)
-        )))
+        .then(|r| {
```

Comment:
Maybe handle timeout error in `LanDiscovery` module? This way we can log it properly.